lib/ansible/modules/windows/win_domain_user.ps1

```diff
@@ -103,13 +193,25 @@ If ($state -eq 'present') {
         $user_obj = Get-ADUser -Identity $username -Properties * @extra_args
     }
 
-    # Set the password if required
-    If ($password -and (($new_user -and $update_password -eq "on_create") -or $update_password -eq "always")) {
-        $secure_password = ConvertTo-SecureString $password -AsPlainText -Force
-        Set-ADAccountPassword -Identity $username -Reset:$true -Confirm:$false -NewPassword $secure_password -WhatIf:$check_mode @extra_args
-        $user_obj = Get-ADUser -Identity $username -Properties * @extra_args
-        $result.password_updated = $true
-        $result.changed = $true
+    If ($password) {
+        # Don't uncecessary check for working credentials.
+        # Set the password if we need to.
+        # For new_users there is also no difference between always and when_changed
+        # so we don't need to differentiate between this two states.
+        If ($new_user -or ($update_password -eq "always")) {
+            $set_new_credentials = $true
+        } elseif ($update_password -eq "when_changed") {
+            $set_new_credentials = -not (Test-Credential -Username $username -Password $password)
```

Comment:
If we are testing the credential of an existing user can we source the username/domain (hopefully the UPN) from the actual AD object instead of relying on the user input?